### PR TITLE
Add DateTime/NaiveDateTime.beginning_of_day and DateTime/NaiveDateTime.end_of_day

### PR DESCRIPTION
**Pull request by thiamsantos**
Addresses https://groups.google.com/g/elixir-lang-core/c/dfAn00VCA9Q

### CODE REVIEW

**Comment by josevalim** [comment]
Hi @kipcole9, could you also please review this PR and let us know your thoughts? Thank you!

**Comment by wojtekmach** [comment]
Is it safe for datetimes with non-utc tz? I remember on clock change some datetimes don’t actually exist and maybe there are time zones when these invalid datetimes would happen to be beginning or end of day?

**Comment by wojtekmach** [comment]
Could you add unit tests for precision? inputs like (…)09:00:00 and (…)09:00:00.123.

**Comment by josevalim** [comment]
Ah, good point. This is timezone aware, which adds a reasonable amount of complexity here.

**Comment by NickNeck** [comment]
Here are two examples for daylight saving changes on a start of a day.
```elixir
~N[2011-04-03 12:00:00]
|> DateTime.from_naive!("Africa/El_Aaiun")
|> Tox.DateTime.beginning_of_day()
=> #DateTime<2011-04-03 01:00:00+01:00 +01 Africa/El_Aaiun>

datetime = DateTime.from_naive!(~N[2020-10-25 12:00:00], "America/Scoresbysund")
=> #DateTime<2020-10-25 12:00:00-01:00 -01 America/Scoresbysund>
Tox.DateTime.beginning_of_day(datetime)
=> #DateTime<2020-10-25 00:00:00+00:00 +00 America/Scoresbysund>
```
And one example for end of the day.
```elixir
~N[1916-06-14 12:00:00]
|> DateTime.from_naive!("Africa/Algiers")
|> Tox.DateTime.end_of_day()
=> #DateTime<1916-06-14 22:59:59.999999+00:00 WET Africa/Algiers>
```

**Comment by kipcole9** [comment]
Not all calendars start the day at `00:00:00`. Julian calendar starts the day at `12:00:00`.  Islamic and Jewish calendars start of day is `sunset`. So there is definitely complexity for some calendars and it can't be assumed that the difference between start of day and end of day is 24 hours (DST already breaks that assumption anyway).

Nevertheless, by definition each day has a beginning and and end so its better that calendars resolve that complexity than developers and I don't see any insurmountable issues with the PR.

The main consideration in my mind, for calendar writers, is that the date components are calendar relative. But time components are always understood to be midnight to midnight.  

For example, in an Islamic calendar the calendar day today is `1444-08-10`. But all time references in Elixir are still midnight to midnight.   

Even though in Istanbul today, where sunset is 18:57pm, I believe the practise would not be to say `1444-08-10 00:00:00` (Gregorian `2023-03-03 18:57+03`) rather it would be to say `1444-08-10 18:57:00+03`. 

That calendar dates are calendar relative but times are always relative to midnight would be worth documenting on the Calendar module.

**Comment by josevalim** [comment]
Thank you @kipcole9! Your expectations pretty much match my mine. My only concern is that I don't know how to tackle this timezone wise. Especially in the sunset cases, where the calendar can only return the sunset if it knows the geographical area it is in.

**Comment by thiamsantos** [comment]
If we only introduce the functions on NaiveDateTime, and avoid dealing with timezones. We could avoid the complexity to make the calendar aware of the timezone?

**Comment by josevalim** [comment]
@thiamsantos it may be the opposite, we can't implement this without timezone information. Or do we have multiple calendars for each timezone in said cases? Thoughts @kipcole9?

**Comment by kipcole9** [comment]
I think the only issue with timezones would be if the timezone change occurs at midnight in which case the situation is the same as any tz conversion - the result can be ambiguous (two occurrences of the same time).  In which case `beginning_of_day` and `end_of_day` would need to return an ambiguous result in such cases.

**Comment by josevalim** [comment]
So my suggestion is to add it only to NaiveDateTime. If someone wants to convert a DateTime to its beginning of day, then they do:

```
datetime
|> NaiveDateTime.beginning_of_day()
|> DateTime.from_naive(datetime.timezone)
```

and the docs for NaiveDateTime can say as much. WDYT?

**Comment by kipcole9** [comment]
I think that’s nice and simple and straight forward for calendar developers too.Sent from my iPhoneOn 4 Mar 2023, at 10:47, José Valim ***@***.***> wrote:﻿
So my suggestion is to add it only to NaiveDateTime. If someone wants to convert a DateTime to its beginning of day, then they do:
datetime
|> NaiveDateTime.beginning_of_day()
|> DateTime.from_naive(datetime.timezone)

and the docs for NaiveDateTime can say as much. WDYT?

**Comment by josevalim** [comment]
@thiamsantos can you please do these changes:

1. Add precision tests (tests, not doctests :))
2. Keep it only on NaiveDateTime
3. Add the docs I mentioned above for DateTime usage

Thank you!

**Comment by thiamsantos** [comment]
@josevalim changes pushed.

Thank you all for the feedback and thoughts :)

File: lib/elixir/lib/calendar/naive_datetime.ex [hunk 1]
```diff
@@ -1169,6 +1169,54 @@ defmodule NaiveDateTime do
...
   end
 
+  @doc """
+  Calculates a `NaiveDateTime` that is the first moment for the given `NaiveDateTime`.
+
+  To calculate the beginning of day of a `DateTime`, call this function, then convert back to a `DateTime`:
+
+      datetime
+      |> NaiveDateTime.beginning_of_day()
+      |> DateTime.from_naive(datetime.timezone)
```
Comment (josevalim) [inline lib/elixir/lib/calendar/naive_datetime.ex:1179]:
```suggestion
      |> DateTime.from_naive(datetime.timezone)

  Note that the beginning of the day may not exist or be ambiguous
  in a given timezone, so you must handle those cases accordingly.
```

File: lib/elixir/lib/calendar/naive_datetime.ex [hunk 2]
```diff
@@ -1169,6 +1169,54 @@ defmodule NaiveDateTime do
...
+  end
+
+  @doc """
+  Calculates a `NaiveDateTime` that is the last moment for the given `NaiveDateTime`.
+
+  To calculate the end of day of a `DateTime`, call this function, then convert back to a `DateTime`:
+
+      datetime
+      |> NaiveDateTime.beginning_of_day()
+      |> DateTime.from_naive(datetime.timezone)
```
Comment (josevalim) [inline lib/elixir/lib/calendar/naive_datetime.ex:1203]:
```suggestion
      |> DateTime.from_naive(datetime.timezone)

  Note that the beginning of the day may not exist or be ambiguous
  in a given timezone, so you must handle those cases accordingly.
```

**Comment by josevalim** [comment]
Btw, can you push similar docs to beginning_of_week and end_of_week?

File: lib/elixir/test/elixir/calendar/naive_datetime_test.exs [hunk 3]
```diff
@@ -348,4 +348,39 @@ defmodule NaiveDateTimeTest do
...
+               month: 1,
+               second: 59,
+               year: 2000
+             }
+
+      assert NaiveDateTime.end_of_day(~N[2000-01-01 23:00:07]) == %NaiveDateTime{
+               calendar: Calendar.ISO,
+               day: 1,
+               hour: 23,
+               microsecond: {999_999, 0},
```
Comment (wojtekmach) [inline lib/elixir/test/elixir/calendar/naive_datetime_test.exs:378]: the fact it's `{999_999, 0}` might lead to surprising behaviour:

```elixir
import ExUnit.Assertions
start = ~N[2023-01-01 00:00:00]
assert NaiveDateTime.end_of_day(start) == 
       start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second)
```

```
** (ExUnit.AssertionError)

Assertion with == failed
code:  assert NaiveDateTime.end_of_day(start) ==
              start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second)
left:  ~N[2023-01-01 23:59:59]
right: ~N[2023-01-01 23:59:59]
```

the difference is:

```elixir
iex> NaiveDateTime.end_of_day(start).microsecond
{999999, 0}

iex> start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second) |> Map.fetch!(:microsecond)
{0, 0}
```

We'd see similar behaviour for different precisions e.g. milliseconds.

I'm honestly not sure what should be the behaviour here.

Comment (wojtekmach) [reply lib/elixir/test/elixir/calendar/naive_datetime_test.exs:378]: one way to look at it is we should be able to write the assertion under question using just sigils:

```elixir
assert NaiveDateTime.end_of_day(~N[2000-01-01 23:00:07]) == ~N[2000-01-01 23:59:59]
```

and the fact we can't is a reason to revisit this?

Comment (josevalim) [reply lib/elixir/test/elixir/calendar/naive_datetime_test.exs:378]: I created a new issue. We should probably truncate it based on the precision: `{(10 ** precision) - 1, precision}`

Comment (thiamsantos) [reply lib/elixir/test/elixir/calendar/naive_datetime_test.exs:378]: I just opened a PR to fix this #12450.

**Comment by thiamsantos** [comment]
> Btw, can you push similar docs to beginning_of_week and end_of_week?

@josevalim What you have in mind for this? As far as I understood the we can incur in problems when dealing with the time that a day starts/ends. As we only have `Date.beginning_of_week` and `Date.end_of_week` we don't need to worry about this.

**Comment by josevalim** [comment]
You are right, we can skip it for now!

**Pull request merged**